pkg/scalers/gitlab_runner_scaler.go

```diff
@@ -0,0 +1,206 @@
+	maxGitlabAPIPageCount = 50
+	// gitlabAPIPerPage is the number of pipelines to query per page.
+	gitlabAPIPerPage = "200"
+)
+
+type gitlabRunnerScaler struct {
+	metricType v2.MetricTargetType
+	metadata   *gitlabRunnerMetadata
+	httpClient *http.Client
+	logger     logr.Logger
+}
+
+type gitlabRunnerMetadata struct {
+	GitLabAPIURL        *url.URL `keda:"name=gitlabAPIURL, order=triggerMetadata, default=https://gitlab.com, optional"`
+	PersonalAccessToken string   `keda:"name=personalAccessToken, order=authParams"`
+	ProjectID           string   `keda:"name=projectID, order=triggerMetadata"`
+
+	TargetPipelineQueueLength int64 `keda:"name=targetPipelineQueueLength, order=triggerMetadata, default=1, optional"`
+	TriggerIndex              int
+}
+
+// NewGitLabRunnerScaler creates a new GitLab Runner Scaler
+func NewGitLabRunnerScaler(config *scalersconfig.ScalerConfig) (Scaler, error) {
+	httpClient := kedautil.CreateHTTPClient(config.GlobalHTTPTimeout, false)
+
+	metricType, err := GetMetricTargetType(config)
+	if err != nil {
+		return nil, fmt.Errorf("error getting scaler metric type: %w", err)
+	}
+
+	meta, err := parseGitLabRunnerMetadata(config)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing GitLab Runner metadata: %w", err)
+	}
+
+	return &gitlabRunnerScaler{
+		metricType: metricType,
+		metadata:   meta,
+		httpClient: httpClient,
+		logger:     InitializeLogger(config, "gitlab_runner_scaler"),
+	}, nil
+}
+
+func parseGitLabRunnerMetadata(config *scalersconfig.ScalerConfig) (*gitlabRunnerMetadata, error) {
+	meta := gitlabRunnerMetadata{}
+
+	meta.TriggerIndex = config.TriggerIndex
+	if err := config.TypedConfig(&meta); err != nil {
+		return nil, fmt.Errorf("error parsing gitlabRunner metadata: %w", err)
+	}
+
+	return &meta, nil
+}
+
+func (s *gitlabRunnerScaler) GetMetricsAndActivity(ctx context.Context, metricName string) ([]external_metrics.ExternalMetricValue, bool, error) {
+	// Get the number of pending, waiting, and running for resource pipelines
+	eg, ctx := errgroup.WithContext(ctx)
+
+	getLen := func(status pipelineStatus, length *int64) func() error {
+		return func() error {
+			uri := constructGitlabAPIPipelinesURL(*s.metadata.GitLabAPIURL, s.metadata.ProjectID, status)
+			var err error
+			*length, err = s.getPipelineQueueLength(ctx, uri)
+			return err
+		}
+	}
+
+	var pendingLen, waitingForResourceLen, runningLen int64
+
+	eg.Go(getLen(pipelinePendingStatus, &pendingLen))
+	eg.Go(getLen(pipelineWaitingForResourceStatus, &waitingForResourceLen))
+	eg.Go(getLen(pipelineRunningStatus, &runningLen))
+
+	err := eg.Wait()
+	if err != nil {
+		s.logger.Error(err, "error getting pipeline queue length")
+		return []external_metrics.ExternalMetricValue{}, false, err
+	}
+
+	queueLen := pendingLen + waitingForResourceLen + runningLen
+
+	metric := GenerateMetricInMili(metricName, float64(queueLen))
+
+	return []external_metrics.ExternalMetricValue{metric}, queueLen >= s.metadata.TargetPipelineQueueLength, nil
```

Comment:
`TargetPipelineQueueLength` shouldn't be used here. It is used for the target, for the activation we need to include another parameter `ActivationTargetPipelineQueueLength` with default value 0 and use it with >
```golang
return []external_metrics.ExternalMetricValue{metric}, queueLen > s.metadata.ActivationTargetPipelineQueueLength, nil
```